### PR TITLE
TypeScript samples

### PR DESCRIPTION
**Pull request by sseiber**
## Purpose

* ...

## Checklist
- [X] I have read the [contribution guidelines](../CONTRIBUTING.md).
- I submitted this PR against the correct branch: 
  - [X] This pull-request is submitted against the `main` branch. 
  - [X] I have merged the latest `main` branch prior to submission and re-merged as needed after I took any feedback.
  - [X] I will squashed my changes into one with a clear description of the change when I complete the PR.

## Does this introduce a breaking change?

```
[ ] Yes
[X] No
```

## Pull Request Type
New TypeScript scenarios

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
See all READMEs. Each scenario describes how to run/test.

## What to Check
TypeScript functionality is the same as other language/scenarios

### CODE REVIEW

File: mqttclients/dotnet/MQTTnet.Client.Extensions/CommandServer.cs
Comment (rido-min) [inline mqttclients/dotnet/MQTTnet.Client.Extensions/CommandServer.cs:52]: is this needed ?

File: mqttclients/ts/mqttjsClientExtensions/package.json
Comment (rido-min) [inline mqttclients/ts/mqttjsClientExtensions/package.json:26]: how can track the issues you pointed out wrt uuid compat with .NET Guids?

File: mqttclients/ts/mqttjsClientExtensions/package.json
Comment (rido-min) [inline mqttclients/ts/mqttjsClientExtensions/package.json:25]: missing tsc as a devDevependency

File: mqttclients/ts/mqttjsClientExtensions/src/geoJsonPoint.ts
Comment (rido-min) [inline mqttclients/ts/mqttjsClientExtensions/src/geoJsonPoint.ts:1]: should we use https://github.com/caseycesari/geojson.js/ ?

File: mqttclients/ts/mqttjsClientExtensions/src/index.ts
Comment (rido-min) [inline mqttclients/ts/mqttjsClientExtensions/src/index.ts:1]: ideally, the extensions (aka helpers) should not know about the samples that consume it. I'm assuming you have this to be able to "build all samples" from the root. If so, this file should be an index.ts in the root, not here.. no?

File: mqttclients/ts/mqttjsClientExtensions/src/utf8JsonSerializer.ts
Comment (rido-min) [inline mqttclients/ts/mqttjsClientExtensions/src/utf8JsonSerializer.ts:11]: do we need to deserialze? I thought we could use the underlying stream to initialize the Buffer

File: scenarios/alert/ts/alert/package.json
Comment (rido-min) [inline scenarios/alert/ts/alert/package.json:1]: seems the alert scenario is still WIP, can we remove it from this PR? and then add it when complete?

File: scenarios/command/README.md
Comment (rido-min) [inline scenarios/command/README.md:279]: we already have this instructions for C, and I believe that we also install those in the devContainer. we should unify these instructions

File: scenarios/command/README.md
Comment (rido-min) [inline scenarios/command/README.md:294]: typo, ts sample.

File: scenarios/command/ts/protoMessages/src/index.ts
Comment (rido-min) [inline scenarios/command/ts/protoMessages/src/index.ts:1]: this file fails to build when the protoc has not been generated. I wish this file could also be produced from the proto

File: scenarios/command/ts/protoMessages/src/proto/unlock_command.proto
Comment (rido-min) [inline scenarios/command/ts/protoMessages/src/proto/unlock_command.proto:1]: should we think of reusing the same proto file across samples?

File: scenarios/command/ts/protoMessages/src/unlockRequestSerializer.ts
Comment (rido-min) [inline scenarios/command/ts/protoMessages/src/unlockRequestSerializer.ts:14]: fromJSON? I thought this was proto

File: scenarios/getting_started/ts/gettingStarted/package.json
Comment (rido-min) [inline scenarios/getting_started/ts/gettingStarted/package.json:18]: can we add tsc to devDependencies?

File: Setup.md
Comment (rido-min) [inline Setup.md:191]: can we install this from package.json?

**Review by rido-min — COMMENTED** [review]
This PR looks amazing !! Thank you very much for your contributions.

I'm adding some minor comments, nothing major or blocking, waiting for your replies before approval.

File: mqttclients/ts/mqttjsClientExtensions/src/geoJsonPoint.ts
Comment (sseiber) [reply mqttclients/ts/mqttjsClientExtensions/src/geoJsonPoint.ts:1]: Not sure. That one is 6-years old. I was keeping it simple, but could go either way.

File: scenarios/command/ts/protoMessages/src/proto/unlock_command.proto
Comment (sseiber) [reply scenarios/command/ts/protoMessages/src/proto/unlock_command.proto:1]: This is a classic conversation in the proto world. But, I'm not sure it's generally talked about across language samples. There may need to be some CI/automation/setup to copy the file to the right sample . Also, the mono-repo I setup may not be able to access it. It's worth further discussion.

File: scenarios/command/ts/protoMessages/src/unlockRequestSerializer.ts
Comment (sseiber) [reply scenarios/command/ts/protoMessages/src/unlockRequestSerializer.ts:14]: I could use some collaboration on the generics here. I may indeed be able to use the underlying stream to combine the UnlockRequest and UnlockResponse serializer. But, I was confused about some of the mechanics here.

File: Setup.md
Comment (sseiber) [reply Setup.md:191]: I don't think so, maybe running a script during install? But, I'm not sure if that is anti-pattern. I took a hint from the C samples which require some pre-installed binaries.

File: mqttclients/ts/mqttjsClientExtensions/src/geoJsonPoint.ts
Comment (rido-min) [reply mqttclients/ts/mqttjsClientExtensions/src/geoJsonPoint.ts:1]: GeoJSON has not changed in 6 years, sometimes is not a bad thing, just means it's _DONE_.

maybe there is a more recent library, this is just to show how to adapt to different messaging formats/patterns.

File: scenarios/command/ts/protoMessages/src/proto/unlock_command.proto
Comment (rido-min) [reply scenarios/command/ts/protoMessages/src/proto/unlock_command.proto:1]: fair enough, I agree I see a lot of manual copies of protobuf files.

File: scenarios/command/ts/protoMessages/src/unlockRequestSerializer.ts
Comment (rido-min) [reply scenarios/command/ts/protoMessages/src/unlockRequestSerializer.ts:14]: have you tried to use [protobuf.js](https://github.com/protobufjs/protobuf.js/) ? it allows to load the protos at runtime and includes a decoder:

here is an example 
https://github.com/iotmodels/iotux-mqtt/blob/main/telemetryProto.js#L58

File: Setup.md
Comment (rido-min) [reply Setup.md]: at least add the command, to save the user one visit to the typescript site :)

File: scenarios/command/ts/protoMessages/src/unlockRequestSerializer.ts
Comment (sseiber) [reply scenarios/command/ts/protoMessages/src/unlockRequestSerializer.ts:14]: I had protobuf.js in from the beginning but swapped it out in favor of a better TypeScript generator. Also, I don't think it was feasible to load the proto at runtime since we needed it in both the initial app, a sub-module of the app, and then in some common code. It also didn't compile the .js into .ts it only provided the .ts types. This required a bunch of pre-post build custom steps. All of that might be worth not having to pre-install the proto compiler. Let's chat about this.

File: Setup.md
Comment (sseiber) [reply Setup.md]: Done.

File: scenarios/getting_started/ts/gettingStarted/package.json
Comment (sseiber) [reply scenarios/getting_started/ts/gettingStarted/package.json:18]: Same as above (mono-repo).

**Review by rido-min — APPROVED** [review]

**Pull request merged**
